daemon/loop.go

```diff
@@ -21,40 +21,74 @@ type LoopVars struct {
 	GitPollInterval      time.Duration
 	RegistryPollInterval time.Duration
 	syncSoon             chan struct{}
-	initSyncSoon         sync.Once
+	pollImagesSoon       chan struct{}
+	initOnce             sync.Once
+}
+
+func (loop *LoopVars) ensureInit() {
+	loop.initOnce.Do(func() {
+		loop.syncSoon = make(chan struct{}, 1)
+		loop.pollImagesSoon = make(chan struct{}, 1)
+	})
 }
 
 // Loop for potentially long-running stuff. This includes running
 // jobs, and looking for new commits.
 
 func (d *Daemon) Loop(stop chan struct{}, wg *sync.WaitGroup, logger log.Logger) {
 	defer wg.Done()
-	pollGit := time.NewTimer(d.GitPollInterval)
+	gitPollTimer := time.NewTimer(d.GitPollInterval)
+	// This can be called any time to push the next git poll out by
+	// d.GitPollInterval. It must _at least_ be called when the timer
+	// fires, so that the polling continues.
 	resetGitPoll := func() {
-		if pollGit != nil {
-			pollGit.Stop()
-			pollGit = time.NewTimer(d.GitPollInterval)
+		if gitPollTimer != nil {
+			gitPollTimer.Stop()
+			gitPollTimer = time.NewTimer(d.GitPollInterval)
+		}
+	}
+
+	// Pull for new commits, then if successful, run the procedure given
+	pullThen := func(k func(logger log.Logger)) {
+		if err := d.Checkout.Pull(); err != nil {
+			logger.Log("operation", "pull", "err", err)
+			return
+		}
+		k(logger)
+	}
+
+	imagePollTimer := time.NewTimer(d.RegistryPollInterval)
+	// Similar to resetGitTimer, this must at lest be called when the
+	// timer fires.
+	resetImagePoll := func() {
+		if imagePollTimer != nil {
+			imagePollTimer.Stop()
+			imagePollTimer = time.NewTimer(d.RegistryPollInterval)
 		}
 	}
 
-	pollImages := time.Tick(d.RegistryPollInterval)
-	// Ask for a sync straight away
+	// Ask for a sync, and to poll images, straight away
 	d.askForSync()
+	d.askForImagePoll()
 	for {
```

Comment:
Yeah it's really two loops, tangled together, and that makes it hard to understand. I'll see if I can detangle them, and if that makes the code clearer.